sql/hive-thriftserver/src/main/java/org/apache/hive/service/auth/HiveAuthFactory.java

```diff
@@ -362,4 +371,34 @@ public static void verifyProxyAccess(String realUser, String proxyUser, String i
     }
   }
 
+  public static boolean needUgiLogin(UserGroupInformation ugi, String principal, String keytab) {
+    return null == ugi || !ugi.hasKerberosCredentials() || !ugi.getUserName().equals(principal) ||
+      !keytab.equals(getKeytabFromUgi());
+  }
+
+  private static String getKeytabFromUgi() {
+    Class<?> clz = UserGroupInformation.class;
+    try {
+      synchronized (clz) {
+        Field field = clz.getDeclaredField("keytabFile");
+        field.setAccessible(true);
+        return (String) field.get(null);
+      }
+    } catch (NoSuchFieldException e) {
+      try {
+        synchronized (clz) {
+          // In Hadoop 3 we don't have "keytabFile" field, instead we should use private method
+          // getKeytab().
+          Method method = clz.getDeclaredMethod("getKeytab");
+          method.setAccessible(true);
+          return (String) method.invoke(UserGroupInformation.getCurrentUser());
```

Comment:
I was not aware of this change, thanks for fixing this !
Let us move both field and method out of getKeytabFromUgi.
```
synchronized(UserGroupInformation.class) {
  if (null != keyTabField) { 
    return keyTabField.get(null); 
  } else {
    return (String) method.invoke(UserGroupInformation.getCurrentUser()) 
  }
}
```

Reply:
What is the purpose of moving both field and method out of this method? I'm not sure is there any difference.

Reply:
This will only be called twice in the initialization stage, so there should not be large overhead.

Reply:
It is called twice right now; but as a util method which can be used in other place, let us not intentionally introduce known inefficiencies.

Reply:
Sure, I will change it.